### ISSUE TITLE
[Winlogbeat] Winlogbeat can't keep up with the EPS value of our Windows Event Collector

### ISSUE THREAD

**Issue by macEar** [elastic/beats#17680]
# Winlogbeat version

```
PS C:\temp\winlogbeat> .\winlogbeat.exe version

winlogbeat version 7.6.2 (amd64), libbeat 7.6.2 [d57bcf8684602e15000d65b75afcd110e2b12b59 built 2020-03-26 05:32:51 +0000 UTC]
```

# Operating System

```
PS C:\temp\winlogbeat> (Get-WMIObject win32_operatingsystem).name

Microsoft Windows Server 2016 Standard...
```

# Bug Info

We have Windows Server used as a Windows Event Collector. With a help of `perfmon.exe` we checked our EPS for WEC. The average value is equal to `1224`, but of course it changes during the day and sometimes is up to 6000 events per second.

Winlogbeat configuration:

```yml
winlogbeat.registry_file: .winlogbeat.yml
winlogbeat.shutdown_timeout: 15s
winlogbeat.event_logs:
  - name: ForwardedEvents
    forwarded: true
    batch_read_size: 1024
    ignore_older: 24h
queue:
  mem:
    events: 65472
    flush.min_events: 0
output.console:
  enabled: true
  codec.json:
    pretty: false
    escape_html: false
path.config: ${path.home}
setup.template.enabled: false
setup.template.name: "winlogbeat-%{[agent.version]}"
setup.template.pattern: "winlogbeat-*"
setup.template.fields: "${path.config}/fields.yml"
setup.template.overwrite: true
setup.template.settings:
  index:
    number_of_shards: 1
    codec: best_compression
    number_of_routing_shards: 30
    number_of_replicas: 0
setup.ilm.enabled: false
logging.level: debug
logging.to_files: true
logging.files:
http.enabled: true
http.host: localhost
http.port: 5066
```

With a help of a simple powershell script we scraped our winlogbeat http endpoint for some metrics to evaluate approximate value of events processed for 30s.

```powershell
$headers = New-Object "System.Collections.Generic.Dictionary[[String],[String]]"
$headers.Add("Accept", 'application/json')
For ($i = 0; $i -lt 5; $i++) { 
$response = Invoke-RestMethod 'http://localhost:5066/stats?pretty' -Headers $headers -Method GET
$result =@{}
$result["№"] = $i.toString()
$result["Total"] = $response[0].libbeat.output.events.total
$result | Format-Table
Start-Sleep -s 30
}
```

The output is as follows:

```
Name                           Value
----                           -----
Total                          241521
№                              0

Name                           Value
----                           -----
Total                          241521
№                              1

Name                           Value
----                           -----
Total                          261888
№                              2

Name                           Value
----                           -----
Total                          261888
№                              3

Name                           Value
----                           -----
Total                          261888
№                              4
```

Quick/rough math:
1. 261888 - 241521 = 20367
Events which has been processed by winlogbeat for 150s (30s * 5)
2. 20367 / 150 = 135,78 
Events per second for winlogbeat
3. Compare it to WEC EPS:

![image](https://user-images.githubusercontent.com/47277788/79095406-82183100-7d84-11ea-8762-11b3dafda2d2.png)

**135,78 is much too less than 778,298**..

We tried to increase `batch_read_size`, but as it is written on your web-site, Windows API returns the error for any value greater than 1024:

```
2020-04-13T12:04:03.658+0700	INFO	service/service_windows.go:101	Attempted to register Windows service handlers, but this is not a service. No action necessary
2020-04-13T12:04:03.660+0700	DEBUG	[eventlog]	eventlog/wineventlog.go:181	WinEventLog[ForwardedEvents] using subscription query=<QueryList>
  <Query Id="0">
    <Select Path="ForwardedEvents">*[System[TimeCreated[timediff(@SystemTime) &lt;= 86400000]]]</Select>
  </Query>
</QueryList>
2020-04-13T12:04:03.661+0700	DEBUG	[winlogbeat]	beater/eventlogger.go:131	EventLog[ForwardedEvents] opened successfully
2020-04-13T12:04:03.661+0700	WARN	eventlog/wineventlog.go:312	WinEventLog[ForwardedEvents] EventHandles returned error The parameter is incorrect.
2020-04-13T12:04:03.661+0700	WARN	beater/eventlogger.go:148	EventLog[ForwardedEvents] Read() error: The parameter is incorrect.
2020-04-13T12:04:03.661+0700	INFO	beater/eventlogger.go:123	EventLog[ForwardedEvents] Stop processing.
```

Is it possible to configure winlogbeat to be able to process all events on time? We are losing a lot of events due to this issue.

**Comment by andrewkroh** [comment]
You can try out the new reader implementation if you like. It should be faster. It's not released yet.

- Docs: https://www.elastic.co/guide/en/beats/winlogbeat/7.7/configuration-winlogbeat-options.html#_event_logs_api
- Download: https://beats-ci.elastic.co/job/elastic+beats+7.7+multijob-package-linux/gcsObjects/

**Comment by macEar** [comment]
@andrewkroh thanks for your reply!

I'll try to upgrade winlogbeat up to 7.7.0 downloaded through the abovementioned link. In `winlogbeat.yml` configuration file I'll add this:

```yml
winlogbeat.event_logs:
  - name: ForwardedEvents
    api: wineventlog-experimental
    forwarded: true
    batch_read_size: 1024
    ignore_older: 24h
```

After that I'll monitor how winlogbeat copes with our volume of events per second. Then I'll write back about results.

**Comment by macEar** [comment]
Tried and from the very begging I stumbled on this issue (below you can see a little part of my debug logs, actually an event published):

```
    "event_data": {
      "UtcTime": "{9f6d0b73-587a-5e8c-0000-00102bd80300}",
      "ProcessGuid": "4376",
      "ProcessId": "C:\\Program Files (x86)\\Kaspersky Lab\\Kaspersky Endpoint Security for Windows\\avp.exe",
      "Image": "C:\\ProgramData\\Kaspersky Lab\\KES\\Temp\\ioc970622A9-2FC6-0D43-B363-0B84605D98ED.ps1",
      "TargetFilename": "2020-04-13 17:14:40.809",
      "RuleName": "2020-04-13 17:14:40.810"
    }
```

As you can notice the values of the fields was mixed up, and this behavior related to almost all other logs. 

If I try to use the experimental API with my previous version of winlogbeat (7.6.2), I get the following error:

```
2020-04-14T02:04:46.477+0700	ERROR	instance/beat.go:933	Exiting: Failed to create new event log. wineventlog-experimental API is not available
```

Are there any possible solutions to make winlogbeat keep up with my windows event forwarding rate for the current version (7.6.2)? What parameters can I try to tune up to make it possible?

**Comment by lsoumille** [comment]
Any news on this issue ? We are also facing some performance issues when forwarding from WEF to ELK

**Comment by raomin** [comment]
Same here, we have the same configuration on the winlogbeat part and it cannot keep up when our WEC reaches more than 2000 eps, which can happen steadily...
CPU/RAM/Disk and Network are pretty low on both side (WEC and logstash).

Are there some configurations schemes that could improve the ingestion of winlogbeat? We tried 2 instances on the WEC with different filters to spread the load and it did not help at all, perfs where even lower...

Now we are in a difficult situation where winlogbeat cannot keep up with the backlog of windows events. Help would be appreciated :)

**Comment by macEar** [comment]
The problem solved. It was resolved after I changed console output to kafka. It wasn't winlogbeat issue, the main reason why I had so low throughput is console output and stdout buffer's size (the system default of io.DEFAULT_BUFFER_SIZE).

**Comment by willemdh** [comment]
Is this new reader available in 7.9?

**Comment by rbr101** [comment]
Hi,

i has also performance problems with reading forwarded event log container on a wec server.
Could not go beyond 2000 eps, resulting in more then 4 hours delayed events .
Then i tried -> api: wineventlog-experimental and performance was not an issue anymore ( 8000 eps) but today i noticed that there were duplicate events with identical windows record id

the option as stated “experimental” is most probable the cause of the event duplication, prob an issue with multithreading or something. Anyway, for use as siem/ sec monitoring not very usefull to have duplicate events. So hope this issue will be resolved in later versions. 

Will have to figur out how to collect and process the eventlogs now ...

My configuration has kafka broker in between, so beat agents output is configured for kafka.

more people having difficulties reading high volume events logs from wec servers ?
